radicale_auth_ldap/__init__.py

```diff
@@ -25,79 +26,126 @@
+        logger.info("  %r is %r", "ldap_filter", self.ldap_filter)
+        logger.info("  %r is %r", "ldap_attribute", self.ldap_attribute)
+        logger.info("  %r is %r", "ldap_binddn", self.ldap_binddn)
+        logger.info("  %r is %r", "ldap_password", self.ldap_password)
+
+    def login(self, login, password):
+        if login == "" or password == "":
+            return ""
+
+        server = ldap3.Server(self.ldap_url, get_info=ldap3.ALL)
+        conn = ldap3.Connection(server=server, user=self.ldap_binddn,
+                                password=self.ldap_password, check_names=True,
+                                lazy=False, raise_exceptions=False)
+        conn.open()
         conn.bind()
 
-        try:
-            self.logger.debug("LDAP whoami: %s" % conn.extend.standard.who_am_i())
-        except Exception as err:
-            self.logger.debug("LDAP error: %s" % err)
-
-        distinguished_name = "%s=%s" % (ATTRIBUTE, ldap3imports.escape_attribute_value(user))
-        self.logger.debug("LDAP bind for %s in base %s" % (distinguished_name, BASE))
-
-        if FILTER:
-            filter_string = "(&(%s)%s)" % (distinguished_name, FILTER)
-        else:
-            filter_string = distinguished_name
-        self.logger.debug("LDAP filter: %s" % filter_string)
-
-        conn.search(search_base=BASE,
-                    search_scope=SCOPE,
-                    search_filter=filter_string,
-                    attributes=[ATTRIBUTE])
-
-        users = conn.response
-
-        if users:
-            user_dn = users[0]['dn']
-            uid = users[0]['attributes'][ATTRIBUTE]
-            self.logger.debug("LDAP user %s (%s) found" % (uid, user_dn))
-            try:
-                conn = ldap3.Connection(SERVER, user_dn, password)
-                conn.bind()
-                self.logger.debug(conn.result)
-                if SUPPORT_EXTENDED:
-                    whoami = conn.extend.standard.who_am_i()
-                    self.logger.debug("LDAP whoami: %s" % whoami)
-                else:
-                    self.logger.debug("LDAP skip extended: call whoami")
-                    whoami = conn.result['result'] == 0
-                if whoami:
-                    self.logger.debug("LDAP bind OK")
-                    return True
-                else:
-                    self.logger.debug("LDAP bind failed")
-                    return False
-            except ldap3.core.exceptions.LDAPInvalidCredentialsResult:
-                self.logger.debug("LDAP invalid credentials")
-            except Exception as err:
-                self.logger.debug("LDAP error %s" % err)
-            return False
-        else:
-            self.logger.debug("LDAP user %s not found" % user)
-            return False
+        if conn.result["result"] != 0:
+            logger.error(conn.result)
+            return ""
+
+        final_search_filter = self.ldap_filter.replace("%username", login)
```

Comment:
As I understand it, login is user-specified, right (I don't know a lot about radicale 3's plugin interface either :laughing: )? In case my assumption is right, this looks very dangerous. I'm not so sure about escaping ldap-queries but I'm almost sure submitting user-controlled content directly inside the query string is a bad idea. Could you have a look at #4 ? According to [ldap3's documentation](https://ldap3.readthedocs.io/en/latest/searches.html?highlight=escape) it seems like `escape_filter_chars()` from `ldap3.utils.conv` would be the right tool for this.